### PR TITLE
feat: lazily evaluate for new random_x template

### PR DESCRIPTION
**Pull request by hansieodendaal**
Description
---
- Changed the merge mining proxy to evaluate RandomX block template changes lazily, similar to the strategy employed for the SHA3 miner. Evaluating this often is very expensive, especially during busy network times with a highly volatile mempool.
- Removed _new template_ spam from logging
- Fixed a bug where the command-line network argument `igor` specified for the merge mining proxy was not recognized in the network-aware 'DomainSeparatedConsensusHasher'
- Updated the merge mining proxy `config.toml` to be in line with core

Motivation and Context
---
- With recent stress tests on `nextnet` that lasted multiple hours, RandomX mining virtually ground to a halt during the stress test, whereas SHAR3 mining was able to continue. 
- One merge mining proxy that was monitored for 45 min went up to 4.5 GB in memory usage. During that time the connected base node only advanced one block, while the merge mining proxy requested 428 new block templates from the base node, all of them unique. This could account for ~1 GB in memory usage as block templates are discarded after 20 minutes.

How Has This Been Tested?
---
- System-level testing during low network activity (_on `igor`_)
- System-level stress testing:
  - coin split stress test (_on `igor`_)
  - one-side stealth transaction stress test (_on `igor`_)

What process can a PR reviewer use to test or verify this change?
---
Code walk-through, system-level testing 

Breaking Changes
---

- [x] None
- [ ] Requires data directory on base node to be deleted
- [ ] Requires hard fork
- [ ] Other - Please specify

### CODE REVIEW

File: applications/minotari_app_grpc/proto/base_node.proto [hunk 1]
```diff
@@ -173,6 +173,7 @@ message NewBlockTemplateResponse {
     NewBlockTemplate new_block_template = 1;
     bool initial_sync_achieved = 3;
     MinerData miner_data = 4;
+    bytes best_previous_block_hash = 5;
```
Comment (SWvheerden) [inline applications/minotari_app_grpc/proto/base_node.proto]: This data is already there, its in the `NewBlockTemplate.NewBlockHeaderTemplate.prev_hash`

File: applications/minotari_merge_mining_proxy/src/block_template_data.rs [hunk 2]
```diff
@@ -83,16 +83,25 @@ impl BlockTemplateRepository {
...
+            );
+            BlockTemplateRepositoryItem::new(block_template)
+        });
+    }
+
+    /// Check if the repository contains a block template with best_previous_block_hash
+    pub async fn contains(&self, current_best_block_hash: FixedHash) -> Option<FinalBlockTemplateData> {
+        let b = self.blocks.read().await;
+        b.values()
+            .find(|item| item.data.template.best_previous_block_hash == current_best_block_hash)
```
Comment (SWvheerden) [inline applications/minotari_merge_mining_proxy/src/block_template_data.rs]:
```suggestion
            .find(|item| item.data.template.new_block_template.header.prev_hash == current_best_block_hash)
```

Comment (SWvheerden) [reply applications/minotari_merge_mining_proxy/src/block_template_data.rs]: Block templates will always be build on top of the latest greatest best block

File: applications/minotari_node/src/grpc/base_node_grpc_server.rs [hunk 3]
```diff
@@ -570,6 +570,14 @@ impl tari_rpc::base_node_server::BaseNode for BaseNodeGrpcServer {
...
                 obscure_error_if_true(report_error_flag, Status::internal(e.to_string()))
             })?;
+        let metadata = handler.get_metadata().await.map_err(|e| {
+            warn!(
+                target: LOG_TARGET,
+                "Could not get metadata for new block template: {}",
+                e.to_string()
+            );
+            obscure_error_if_true(report_error_flag, Status::internal(e.to_string()))
+        })?;
```
Comment (SWvheerden) [inline applications/minotari_node/src/grpc/base_node_grpc_server.rs]:
```suggestion

```

File: applications/minotari_node/src/grpc/base_node_grpc_server.rs [hunk 4]
```diff
@@ -585,8 +593,8 @@ impl tari_rpc::base_node_server::BaseNode for BaseNodeGrpcServer {
                     .try_into()
                     .map_err(|e| obscure_error_if_true(report_error_flag, Status::internal(e)))?,
             ),
-
             initial_sync_achieved: status_watch.borrow().bootstrapped,
+            best_previous_block_hash: metadata.best_block_hash().to_vec(),
```
Comment (SWvheerden) [inline applications/minotari_node/src/grpc/base_node_grpc_server.rs]:
```suggestion
```

File: common/src/configuration/utils.rs [hunk 5]
```diff
@@ -82,7 +82,11 @@ pub fn load_configuration_with_overrides<P: AsRef<Path>, TOverride: ConfigOverri
     };
 
     info!(target: LOG_TARGET, "Configuration file loaded.");
-    let overrides = overrides.get_config_property_overrides(network);
+    let overrides = overrides.get_config_property_overrides(&mut network);
+    // Set the network environment variable according to the chosen network (this is to ensure that
+    // `get_current_or_user_setting_or_default()` uses the environment variable if the static network is not set)
+    std::env::set_var("TARI_NETWORK", network.as_key_str());
```
Comment (SWvheerden) [inline common/src/configuration/utils.rs]: I dont like this.

All this does is going to cause much more confusing, we are changing and setting the network in the configuration loading function?

And to make it worse, `TARI_NETWORK != network` its two different fields

File: common/src/configuration/network.rs [hunk 6]
```diff
@@ -49,10 +49,16 @@ pub enum Network {
...
         match CURRENT_NETWORK.get() {
             Some(&network) => network,
-            None => Network::default(),
+            None => {
+                // Check to see if the network has been set by the environment, otherwise use the default
+                match std::env::var("TARI_NETWORK") {
+                    Ok(network) => Network::from_str(network.as_str()).unwrap_or(Network::default()),
+                    Err(_) => Network::default(),
+                }
+            },
```
Comment (SWvheerden) [inline common/src/configuration/network.rs:61]: I am not a fan of this change. 
This is just going to cause more confusing.

Comment (SWvheerden) [reply common/src/configuration/network.rs:61]: okay this works as intended

File: common/src/configuration/utils.rs (hunk 5 above)
Comment (hansieodendaal) [reply common/src/configuration/utils.rs]: Please see improved implementation

File: common/src/configuration/network.rs (hunk 6 above)
Comment (hansieodendaal) [reply common/src/configuration/network.rs:61]: This will enable the code to recognize the user selected network, instead of just using the default

**Review by SWvheerden — APPROVED** [review]

**Pull request merged**